packages/orm/src/FlushProxy.ts

```diff
@@ -0,0 +1,120 @@
+import { OneToManyCollection } from "./collections/OneToManyCollection";
+import { ManyToManyCollection } from "./collections/ManyToManyCollection";
+import { ManyToOneReference } from "./collections/ManyToOneReference";
+import { Entity, EntityManager } from "./EntityManager";
+import * as util from "util";
+
+// These proxies are passed into hooks during a flush to enable changes mid flush
+
+type FlushProxy<T> = T & { __flushSecret: number; __proxyTarget: T };
+type Loader = OneToManyCollection<any, any> | ManyToManyCollection<any, any> | ManyToOneReference<any, any, any>;
+
+export function isFlushProxy<T>(objectOrProxy: T): objectOrProxy is FlushProxy<T> {
+  return util.types.isProxy(objectOrProxy) && (objectOrProxy as any).__flushSecret !== undefined;
+}
+
+export function getTargetFromMaybeFlushProxy<T>(objectOrProxy: T | FlushProxy<T>): T {
+  return isFlushProxy(objectOrProxy) ? objectOrProxy.__proxyTarget : objectOrProxy;
+}
+
+function proxyLoaderResult(result: any, flushSecret: number) {
+  if (Array.isArray(result)) {
+    return result.map((entity) => proxyEntity(entity, flushSecret));
+  } else if (typeof result === "object") {
+    return proxyEntity(result, flushSecret);
+  }
+
+  return result;
+}
+
+function createFlushProxy<T extends object>(
+  target: T,
+  flushSecret: number,
+  block: (target: T, property: PropertyKey, receiver: any) => any,
+): FlushProxy<T> {
+  if (isFlushProxy(target)) {
+    return target;
+  }
+
+  return new Proxy(target, {
+    get(target, property, receiver) {
+      if (property === "__flushSecret") {
+        return flushSecret;
+      }
+
+      if (property === "__proxyTarget") {
+        return target;
+      }
+
+      return block(target, property, receiver);
+    },
+  }) as FlushProxy<T>;
+}
+
+function proxyLoader(loader: Loader, entity: Entity, flushSecret: number): FlushProxy<Loader> {
+  return createFlushProxy(loader, flushSecret, (loader, property, receiver) => {
+    if (property === "load") {
+      return function () {
+        return loader.load().then((result) => proxyLoaderResult(result, flushSecret));
+      };
+    }
+
+    if (property === "get") {
+      return proxyLoaderResult(loader.get, flushSecret);
+    }
+
+    if (property === "entity") {
+      return proxyEntity(entity, flushSecret);
+    }
+
+    return Reflect.get(loader, property, receiver);
+  });
+}
+
+export function proxyEntityManager(em: EntityManager, flushSecret: number): FlushProxy<EntityManager> {
+  return createFlushProxy(em, flushSecret, (em, property, receiver) => {
+    const value = Reflect.get(em, property, receiver);
+
+    // async functions that return entities
+    if (
+      typeof property === "string" &&
+      ["load", "loadAll", "loadFromQuery", "find", "findOne", "findOneOrFail", "findOrCreate"].includes(property)
+    ) {
+      return function (...args: any[]) {
+        return value.apply(receiver, args).then((result: any) => proxyLoaderResult(result, flushSecret));
```

Comment:
I.e. here instead of checking method names, go ahead and let `Reflect.get` do it's thing, it value is a function, return a new function that calls that 1st function, and if the return value is a promise, `then`s it, and then just returns a new proxy...

Reply:
The problem here isn't the promise, it's the eventually result.  Those are separate objects that aren't going to get proxied naturally, so we need to do it ourselves.

Reply:
Disregard the previous, i think i understand what you're saying better now.

Reply:
>  The entire entity chain needs to be composed of proxies or things won't work

Right, I didn't mean stop doing proxies, if anything I meant "proxy all the things" with less checks for "is the method we're going x or y or z"...

Like, if you could/as a starting pointer, have a handler like:

```
handler = get(target, prop) {
  // trap __sessionKey and __proxyTarget 
  value = Reflect.get(target, prop)
  return new Proxy(value, handler);
}
``` 

Naively, wouldn't that work? Or at least as a starting point?

Reply:
Maybe if this weren't javascript it would work 🤣  Proxy behaves weirdly with some data types (eg Date) and won't work at all with primitives.  I think get what you're after though.  

I've stripped out the `proxyEntity`/`proxyEntityManager`/`proxyLoader` methods locally and replaced them with the much more generic:

```typescript
function maybeProxy<T extends any>(object: T, flushSecret: number): any {
  if (typeof object === "function" && !/^class\s/.test(object.toString())) {
    return function (this: any, ...args: any[]) {
      return maybeProxy(object.apply(this, args), flushSecret);
    };
  } else if (typeof object === "object" && object !== null) {
    if (object instanceof Promise) {
      return object.then((result) => maybeProxy(result, flushSecret));
    } else if (Array.isArray(object)) {
      return object.map((element) => maybeProxy(element, flushSecret));
    } else if (
      object instanceof EntityManager ||
      object instanceof BaseEntity ||
      object instanceof OneToManyCollection ||
      object instanceof ManyToManyCollection ||
      object instanceof ManyToOneReference
    ) {
      return createFlushProxy(object, flushSecret);
    }
  }

  return object;
}

export function createFlushProxy<T extends object>(target: T, flushSecret: number): FlushProxy<T> {
  if (isFlushProxy(target)) {
    return target;
  }

  return new Proxy(target, {
    get(target, property, receiver) {
      if (property === "__flushSecret") {
        return flushSecret;
      }

      if (property === "__proxyTarget") {
        return target;
      }

      // knex seems to really dislike being put through maybeProxy
      if (property === "knex") {
        return (target as any).knex;
      }

      return maybeProxy(Reflect.get(target, property, receiver), flushSecret);
    },
  }) as FlushProxy<T>;
}
```

This is failing on 4 tests, all related to hooks.  It's tricky to debug.

Reply:
Ah yeah, sorry about the overlapping comment, I had it drafted and didn't see your reply. 

Yep, that's exactly what I'm wondering if it would work...

Why does your `maybeProxy` not have a `new Proxy` call it in? Oh I see, it's only for that last "else if"... this is going to sound super-naive, but what if `maybeProxy` was not `maybeProxy` and instead `alwaysProxy`, i.e change `return maybeProxy(Reflect.get...` to `return createFlushProxy(Reflect.get)`...

Reply:
How about a `FlushProxy.test.ts` that creates an EM/entity, manually wraps the proxy, and then starts various call chains of `entity.__flushSecret`, `author.books.load()...[0].__flushSecret`, `author.publisher.get.__flushSecret`, etc. that you want to all expect to work.

Reply:
...the test might have to live in the integration-tests to be able to use the domain objects. But that's probably fine.

Reply:
Many things in javascript cannot be proxied (eg primitives, it's an immediate error), behave unpredictably (eg, Date), or break in other ways (eg, class constructors).

Reply:
>  Proxy behaves weirdly with some data types (eg Date) and won't work at all with primitives

Yeah, it's annoying that it's not easier to differentiate primitives/value-ish-objects from classes / object literals...

Reply:
I poked at this just enough locally to see that, right, you pretty quickly hit things like ` Proxy behaves weirdly with some data types (eg Date) and won't work at all with primitives`...

Reply:
Huh. Wonder if this is applicable:

https://github.com/tc39/ecma262/issues/1114
https://tvcutsem.github.io/js-membranes
https://github.com/ajvincent/es-membrane

Reply:
After thinking on it, yeah, since proxies seem to have more boundary cases than expected, and the bespoke-proxy-per-loader/entity/EM was working/passing tests, I'm good with going with that approach. I might ask for that `FlushProxy.test.ts` to make future iteration on this easier, but otherwise this looks great.

(I spent ~15-20m looking at the membrane stuff and it seems non-trivial? Was hoping it'd be like ~10 lines of code to drop it in and do what we want, but doesn't seem like it...)

Reply:
https://github.com/salesforce/observable-membrane looks like a much nicer membrane API...

https://github.com/salesforce/observable-membrane/pull/48 is interesting, there is a comment from jdalton, who created lodash, and the github popup says hes as salesforce as a UI security engineer. Afaiu a usecase of membranes is to provide secure/protected views of an object graph to unprotected/untrusted code, so I assume that is what their library is mainly for.

Reply:
Huh, I scanned the source code of the `observable-membrane` project and it's surprisingly short for what it asserts to do...

Reply:
"This membrane implementation is tailored for observable objects, which are objects with the prototype chain set to Object.prototype or null. Any other object is not wrapped in a Proxy by design." ah shoot from `observable-membrane`...

Reply:
https://github.com/salesforce/observable-membrane/issues/50